### PR TITLE
Request for comment (please don't merge yet)

### PR DESCRIPTION
**Pull request by javaarchive**
# Details I'd appreciate feedback on
* Are some softwares covered out of scope, for example my firefox tangent. Firefox however is installed by default on Ubuntu 22.04. 
* Should mostly windows targeted tools be removed or moved to their own section (might be out of scope as well).
* For NOTES.md would you appreciate a note directing contributing users to write their notes at the end of the file or at a header so they don't conflict with your existing bullet trees?
# Up for discussion
* How to reference content from other sections, for example firefox user.js is the same as prefs.js in `file-overwrite` and can be included in `file-append` and `file-create`. For now I just write "see the section in file-overwrite", maybe obsidian has a fancy link to header thing?
* Possibly out of scope: New area for file read because you can do some cool stuff `cat /proc/1/environ` to read environment of a process by id. 
* How would we link relevant sources for information. For now I'm just putting a line break and then a link that says "Source". 
* Should I leave out ideas I haven't tested yet into some place gitignorable instead of just putting them with the "(untested)" in the header. 
* Better way to write comments to those contributing that isn't prefixing the sentence with "TODO: "
* Similar to sources question, preferably how would I embed a demonstration of a technique. Would the demonstration be kept in a separate file or would I just write it directly in the section. If I prefer to show by video or asciinema, how would you prefer I format such a link. 
# Questions
* Why not try discussions instead of issues, since the structure seems more fitting? 

I apologize if there's anything confusing in the above, I'm writing this quite a bit late at night :) so bear with me

### CODE REVIEW

**Comment by LiveOverflow** [comment]
- I would not consider any software or OS out of scope. However I do think we should focus and prioritize the most useful impacts. For example in a default installation "delete arbitrary file" (probably) does not lead to any privilege escalation. However if we identify software that would make this exploitable, then it's worth documenting. In contrast, "arbitrary file write as root" has so many ways of exploitation on a default installation, that investigating optional software doesn't really help us. How Firefox fits into this, I'm not sure!
- I have never participated or managed an open source project. So I have no idea what should be enforced. At the moment I think the commit volume will be low and we don't really care about "version and change history" like a software does. So I think we could ignore that for now?
- I think Windows is definitely an important OS to research, just not sure how to best do it. Right now we collect everything in a single `README.md` but we could probably split them up once more content is added, and the README simply points at the other .md files. So we have OS specific `WINDOWS.md`, `Ubuntu2204.md`, ... etc.
- I used the NOTES.md to basically document "TODOs" which anybody else could also work on. I kind of like the prefix "TODO:" because somebody could easily grep for open TODOs to work on. Also I don't think the volume is so large that we would need etiquette how to work on the NOTES.md. Maybe we could create new folder "TODO", and in there we collect `NOTES.md` files and anybody can just create a new file. For example if sb. focuses on windows stuff they can just make a `WINDOWS.md`.

- You can just reference other content with markdown links. They should also work in obsidian. 
- If the same file applies to multiple abilities, I think a simple link is enough as you said. "see also: "
- Good idea about the "reads". Documenting interesting file locations to read from could be useful! But probably a simple "list" is enough which could be easily thrown into a script to test which files exist on a target?
- Linking to sources etc, I also wouldn't overthink it. As you said, just add a line and the source link.
- I like ideas to be somehow part of the repo, because it can be like TODOs for people who would want to contribute. Also then it's part of the git history and if somebody really needs to look deeper in one case they might be able to find old notes about it. But maybe as mentioned above we put that stuff in a TODO folder or something?
- The goal for me is to have a resource useable by a professional pentester who found a file write or file delete, and wants to know how to escalate it. Listing a simple example file with some comments what it does should be enough ([like we did on stream](https://github.com/LiveOverflow/security-research/tree/main/arbitrary-file-write/file-create#clean-file)). Asciinema or video is maybe a bit overkill, but sure! if we have it, or even a timestamp to the VOD where we researched it, that can also be useful. But I think a simple link in the comment section should be enough?

- I have no clue about these github features, I enabled discussions now. Maybe we could also use discussions instead of the TODO files? what do you think? Any experience with the feature?

Also feel free to move this over into discussions, if you know how to best organize it :D

File: arbitrary-file-write/file-append/README.md [hunk 1]
```diff
@@ -2,4 +2,49 @@
 * as root or user you can append data to a file and ...
 	* fully control content
 	* partially control content (eg. parts of a logfile)
-	* none (empty file)
\ No newline at end of file
+	* none (empty file)
+
+### sshd
+Looking at a mostly-default sshd config file most things are commented out so anything you append will take precedence if there is not something already before it (TODO: add sources for order of parsing).
```
Comment (LiveOverflow) [inline arbitrary-file-write/file-append/README.md:8]: this needs confirmation. What takes precedence? Can a second config overwrite the old one or not?

File: arbitrary-file-write/file-append/README.md [hunk 2]
```diff
@@ -2,4 +2,49 @@
...
+
+# Example of overriding settings on a per-user basis
+#Match User anoncvs
+#	X11Forwarding no
+#	AllowTcpForwarding no
+#	PermitTTY no
+#	ForceCommand cvs server
+```
+
+#### Messy file ending (user specific and all users)
```
Comment (LiveOverflow) [inline arbitrary-file-write/file-append/README.md:26]: I'm looking at the clean file and the messy file but I cannot tell what is "messy" about it. does it allow binary data? How does it react to just random text? 

The messy file should contain all the possible garbage that still keep it valid

File: arbitrary-file-write/file-create/README.md [hunk 3]
```diff
@@ -49,4 +49,21 @@ asdas
 
 [we]! can have symbols and garbage at the end
 <00><ff> binary also ok here
-```
\ No newline at end of file
+```
+### Shell Profiles
+#### Login Shells
+`--login` flag. 
+* `/etc/profile` executed for all users
```
Comment (LiveOverflow) [inline arbitrary-file-write/file-create/README.md:56]: Does this file not already exist? Should this not be moved over to the `file-overwrite` category.
Also we need examples how it reacts to script errors, binary data, ... 
We need examples how to execute an actual command.

File: arbitrary-file-write/file-create/README.md [hunk 4]
```diff
@@ -49,4 +49,21 @@ asdas
...
+* First found of `~/.bash_profile`, `~/.bash_login`, and `~/.profile` is then executed
```
Comment (LiveOverflow) [inline arbitrary-file-write/file-create/README.md:57]: I guess these files might not exist. in that case it's relevant to this `file-create` section. However we are also missing examples, messy files, etc.

File: arbitrary-file-write/file-create/README.md [hunk 5]
```diff
@@ -49,4 +49,21 @@ asdas
...
+#### Non-login shell
+* `/etc/bash.bashrc` executed for all users 
+* `~/.bashrc` for user specific.  
+
+### Eicar antimalware testfile
```
Comment (LiveOverflow) [inline arbitrary-file-write/file-create/README.md:62]: Generally I don't think it's really useful to an attacker. It's a funny "trick" to trip up antivirus, but it doesn't help you to elevate priviledges.

HOWEVER, this got me thinking. We could investigate if there are antivir that by default delete a file. and maybe this could be used to force delete log file etc. This would be useful to an attacker. But definitely requires more investigation and documentation which antivir this affects.

File: arbitrary-file-write/file-create/README.md [hunk 6]
```diff
@@ -49,4 +49,21 @@ asdas
...
+You can try to obtain it [here](https://www.eicar.org/download-anti-malware-testfile/). The most it'll do is trigger deletion by antivirus especially those with real time detection but an be helpful for triggering weird errors in some applications.  
+
+### git hooks
+One may be able to sneak in a [git hook](https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks#_git_hooks) in the `.git` of a git repo to execute any code of their choosing when a git action happens, however this requires the file to marked executable.
```
Comment (LiveOverflow) [inline arbitrary-file-write/file-create/README.md:66]: good idea, definitely interesting for a server as well, where a deploy script might execute a git pull. But again, we need actual examples, limitations, messy files, etc

File: arbitrary-file-write/file-delete/README.md [hunk 7]
```diff
@@ -8,4 +8,14 @@
 * `rm /etc/group`
 	* system broken. tested: `sudo, login, su`
 * `rm /etc/sudoers`
-	* only `sudo` broken (also tested empty file)
\ No newline at end of file
+	* only `sudo` broken (also tested empty file)
+* `rm /etc/fstab`
+    * say goodbye to your system ig. most likely it'll boot to some rescue mode where you'll have to log in over a display or serial.
```
Comment (LiveOverflow) [inline arbitrary-file-write/file-delete/README.md:13]: "most likely" is not confirmed. We should test this or be more clear about this being unconfirmed.

File: arbitrary-file-write/file-delete/README.md [hunk 8]
```diff
@@ -8,4 +8,14 @@
...
 	* system broken. tested: `sudo, login, su`
 * `rm /etc/sudoers`
-	* only `sudo` broken (also tested empty file)
\ No newline at end of file
+	* only `sudo` broken (also tested empty file)
+* `rm /etc/fstab`
+    * say goodbye to your system ig. most likely it'll boot to some rescue mode where you'll have to log in over a display or serial. 
+
+### shell profile fallbacks
+See shell profile section in file-create, deleting one file will cause the shell to look for another file. Use case: the system sets the path in one of the shell profiles so it uses an up to date version of some programming language, but we bypass that by deleting the profile.
```
Comment (LiveOverflow) [inline arbitrary-file-write/file-delete/README.md:16]: we also need actual examples. "copy-n-paste" ready for somebody who wants to do it

File: arbitrary-file-write/file-delete/README.md [hunk 9]
```diff
@@ -8,4 +8,14 @@
...
\ No newline at end of file
+	* only `sudo` broken (also tested empty file)
+* `rm /etc/fstab`
+    * say goodbye to your system ig. most likely it'll boot to some rescue mode where you'll have to log in over a display or serial. 
+
+### shell profile fallbacks
+See shell profile section in file-create, deleting one file will cause the shell to look for another file. Use case: the system sets the path in one of the shell profiles so it uses an up to date version of some programming language, but we bypass that by deleting the profile. 
+
+### iptables neturalized (untested)
+Some systems (TODO: test which systems do), persist iptables to a specific file so that rules can be reapplied on reboot. Theoritically one could delete such a file so that no rules are applied on reboot.
```
Comment (LiveOverflow) [inline arbitrary-file-write/file-delete/README.md:19]: really good idea! definitely worth looking into. But maybe we should move these ideas into a discussion, TODO file or issue?

File: arbitrary-file-write/file-overwrite/README.md [hunk 10]
```diff
@@ -3,4 +3,86 @@
 * as root or user you can overwrite a file and ...
 	* fully control content
 	* partially control content (eg. parts of a logfile)
-	* none (empty file)
\ No newline at end of file
+	* none (empty file)
+
+### pam
+If we know enough about the target os we can overwrite their user authentication system logic with a backdoor. Here's an [example](https://infosecwriteups.com/creating-a-backdoor-in-pam-in-5-line-of-code-e23e99579cd9) that returns a successful login for a specific hardcoded password.
```
Comment (LiveOverflow) [inline arbitrary-file-write/file-overwrite/README.md:9]: I think it would be useful to have the "stackoverflow" policy, where when you reference a blog, make sure to quote all the relevant pieces here as well, in case the link goes dead.

**Comment by lsalvadore** [comment]
@javaarchive: My comment is not related to your commits, but I just wanted to suggest you to use the "Convert to draft" pull request feature that you can find  in the reviewers section. This way:
- you don't need to add "(please don't merge yet)" to your pull requests titles;
- it is clearer when your pull request is ready: you just need to mark it as ready.

See https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests#draft-pull-requests for further information.

Maybe you already knew about this feature, but since I have discovered it only recently, I thought that others might ignore it as well.
